src/fiat_shamir/poseidon/constraints.rs

```diff
@@ -0,0 +1,325 @@
+        {
+            self.apply_ark(&mut state, i as usize)?;
+            self.apply_s_box(&mut state, true)?;
+            self.apply_mds(&mut state)?;
+        }
+
+        self.state = state;
+        Ok(())
+    }
+
+    #[tracing::instrument(target = "r1cs", skip(self))]
+    fn absorb_internal(
+        &mut self,
+        rate_start_index: usize,
+        elements: &[FpVar<F>],
+    ) -> Result<(), SynthesisError> {
+        // if we can finish in this call
+        if rate_start_index + elements.len() <= self.rate {
+            for (i, element) in elements.iter().enumerate() {
+                self.state[i + rate_start_index] += element;
+            }
+            self.mode = PoseidonSpongeState::Absorbing {
+                next_absorb_index: rate_start_index + elements.len(),
+            };
+
+            return Ok(());
+        }
+        // otherwise absorb (rate - rate_start_index) elements
+        let num_elements_absorbed = self.rate - rate_start_index;
+        for (i, element) in elements.iter().enumerate().take(num_elements_absorbed) {
+            self.state[i + rate_start_index] += element;
+        }
+        self.permute()?;
+        // Tail recurse, with the input elements being truncated by num elements absorbed
+        self.absorb_internal(0, &elements[num_elements_absorbed..])
+    }
+
+    // Squeeze |output| many elements. This does not end in a squeeze
+    #[tracing::instrument(target = "r1cs", skip(self))]
+    fn squeeze_internal(
+        &mut self,
+        rate_start_index: usize,
+        output: &mut [FpVar<F>],
+    ) -> Result<(), SynthesisError> {
+        // if we can finish in this call
+        if rate_start_index + output.len() <= self.rate {
+            output
+                .clone_from_slice(&self.state[rate_start_index..(output.len() + rate_start_index)]);
+            self.mode = PoseidonSpongeState::Squeezing {
+                next_squeeze_index: rate_start_index + output.len(),
+            };
+            return Ok(());
+        }
+        // otherwise squeeze (rate - rate_start_index) elements
+        let num_elements_squeezed = self.rate - rate_start_index;
+        output[..num_elements_squeezed].clone_from_slice(
+            &self.state[rate_start_index..(num_elements_squeezed + rate_start_index)],
+        );
+
+        // Unless we are done with squeezing in this call, permute.
+        if output.len() != self.rate {
+            self.permute()?;
+        }
+        // Tail recurse, with the correct change to indices in output happening due to changing the slice
+        self.squeeze_internal(0, &mut output[num_elements_squeezed..])
+    }
+}
+
+impl<F: PrimeField> AlgebraicSpongeVar<F, PoseidonSponge<F>> for PoseidonSpongeVar<F> {
+    fn new(cs: ConstraintSystemRef<F>) -> Self {
+        // Requires F to be Alt_Bn128Fr
+        let full_rounds = 8;
+        let partial_rounds = 29;
+        let alpha = 17;
```

Comment:
TODO: A remark is likely needed to say that these parameters may not work for all the curves. In the future, the sponge would be supposed to find good parameters for different curves,